### DISCUSSION TITLE
Accessing Microsoft Planetary Computer asset data

### DISCUSSION THREAD

**Discussion by giswqs** [developmentseed/titiler#331] [Q&A] [answered]
I am trying to use Titiler to access Microsoft Planetary Computer asset data on Azure Blob Storage (see [notebook example](https://github.com/microsoft/PlanetaryComputerExamples/blob/main/datasets/landsat-8-c2-l2/landsat-8-c2-l2-example.ipynb)). My goal is to load the asset data to an interactive map created using [leafmap](https://github.com/giswqs/leafmap). However, it seems Titiler cannot access the URLs signed by the planetary-computer package. See the screenshot below. Here is the [link](https://gist.github.com/giswqs/0d5d7711304caf1e6b5429be89f8dcb1) to the notebook. Any advice? Thanks. @vincentsarago @lossyrob

![image](https://user-images.githubusercontent.com/5016453/122805620-52718c80-d297-11eb-8e89-c07597d2154c.png)

**Comment by vincentsarago** [comment, marked as answer]
I think this is because it's a signed url. TiTiler can't support this because it cannot recognize the url parameters (for the signed url) and the url parameter for titiler itself. 

e.g here is what I'm getting with my titiler instance 

```
curl https://api.cogeo.xyz/cog/info?url=https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2020/046/027/LC08_L2SP_046027_20200908_20200919_02_T1/LC08_L2SP_046027_20200908_20200919_02_T1_SR_B4.TIF?st=2021-06-20T19%3A56%3A47Z&se=2021-06-21T20%3A41%3A47Z&sp=rl&sv=2020-02-10&sr=c&skoid=c85c15d6-d1ae-42d4-af60-e2ca0f81359b&sktid=72f988bf-86f1-41af-91ab-2d7cd011db47&skt=2021-06-21T18%3A56%3A47Z&ske=2021-06-22T19%3A56%3A47Z&sks=b&skv=2020-02-10&sig=HI8FeCXdIo44gAEAnlmTcM1XbRboWCIka4LUQmXjUzU%3D

>>> {
  "detail": "'/vsicurl/https://landsateuwest.blob.core.windows.net/landsat-c2/level-2/standard/oli-tirs/2020/046/027/LC08_L2SP_046027_20200908_20200919_02_T1/LC08_L2SP_046027_20200908_20200919_02_T1_SR_B4.TIF?st=2021-06-20T19:56:47Z' does not exist in the file system, and is not recognized as a supported dataset name."
}

```
see the rasterio error is missing some parameter in the URL. 

That's said, in your screenshot, rasterio seems to give back the full url but I wonder if there are some encoding issue.

Sadly there is no easy fix. Usually when we work with signed url we setup a custom tiler where we pass the signed url in a base64 form and then use a custom `url` parameter that decodes the url. I could add an example in TiTiler but I don't think we want to support this by default

**Reply by giswqs** [reply]
@vincentsarago Many thanks for looking into this. It would be great if you can add a notebook example for using signed URLs. I don't expect Titiler to support this by default, but it would be great to see some potential workarounds. Thank you.

**Reply by kylebarron** [reply]
> Usually when we work with signed url we setup a custom tiler where we pass the signed url in a base64

> It would be great if you can add a notebook example for using signed URLs

Seems like it would be hard to put into a notebook example because the workaround really needs to happen on the backend in the tiling code itself, not the notebook. I.e. you'd want the tiler to do something like

```py
# check if image is hosted on Azure, and if so, sign url
if 'windows.net' in url:
	url = pc.sign(url)
```

I agree this is too specific to do by default, but it may be worthwhile to add an optional "url callback argument" to be run by the tiler. That way someone deploying a titiler instance can easily add a function with logic such as the above without needing to subclass all the factory methods.

**Reply by vincentsarago** [reply]
feel free to comment/edit on https://github.com/developmentseed/titiler/blob/signedURLExample/docs/examples/code/working_with_signed_urls.md
